## PGA Draftkings Notebook
Use [PGA Website](https://www.pgatour.com/tournaments/schedule.html) to look up tournament info and fill out first USER INPUT block below. (Fetch/XHR from Network tab, Paylod sub-tab)

Looks like the tournament ID is also in the address bar.

### User Input

In [11]:
# === USER INPUTS ===
# Old Tournament
old_tournament_name = "The Open Championship"
tournament_date = "7/19/2026"  # Ending date of tournament
old_course = "Royal Birkdale GC"
tournament_id = "R2026100"  # Tournament ID from the PGA Tour API - Also visible in the URL of the tournament page

# New Tournament
new_tournament_name = "3M Open"
new_ending_date = "7/26/2026"
new_course = "TPC Twin Cities"
new_season = 2026

# Test mode: set True to dry-run the workflow against a past week's config.
# Skips the odds DB write and relaxes the stale-config guard.
test_mode = False


# === LIBRARIES AND VARIABLES ===
# Import necessary libraries
import requests
import pandas as pd
from datetime import datetime
import sqlite3 as sql
import numpy as np
from numpy import nan
import os
import importlib
import utils.db_utils
from utils.db_utils import TOURNAMENT_NAME_MAP, PLAYER_NAME_MAP

# === TOURNAMENT CONFIG ===
tournament_config = {
    "old": {
        "name": old_tournament_name,
        "date": tournament_date,
        "course": old_course,
        "id": tournament_id
    },
    "new": {
        "name": new_tournament_name,
        "course": new_course,
        "season": new_season,
        "ending_date": pd.to_datetime(new_ending_date, format="%m/%d/%Y")
    }
}


### New Course / Tournament History Check

Exact-match history for `new_tournament_name` / `new_course` **before** running the rest of the notebook. Course-history joins are exact-string matches, so a tournament whose past rows show a differently spelled course (or an unexpectedly empty table) means fix the spelling first.

In [12]:
# === NEW COURSE / TOURNAMENT HISTORY CHECK ===
# Prior events exactly matching new_tournament_name or new_course. A misspelled
# course shows up as tournament rows whose COURSE differs from new_course; an
# empty table for an event that should have history means check the spelling.
with sql.connect("data/golf.db") as _conn:
    _events = pd.read_sql(
        "SELECT DISTINCT SEASON, ENDING_DATE, TOURNAMENT, COURSE FROM tournaments",
        _conn, parse_dates=["ENDING_DATE"])

_hist = (_events[(_events["TOURNAMENT"] == new_tournament_name) |
                 (_events["COURSE"] == new_course)]
         .sort_values("ENDING_DATE", ascending=False)
         .reset_index(drop=True))
print(f"new_tournament_name = {new_tournament_name!r}: "
      f"{(_hist['TOURNAMENT'] == new_tournament_name).sum()} prior events")
print(f"new_course = {new_course!r}: {(_hist['COURSE'] == new_course).sum()} prior events")
if _hist.empty:
    print("No exact matches on either name -- new event, or check spelling.")
_hist

new_tournament_name = '3M Open': 7 prior events
new_course = 'TPC Twin Cities': 7 prior events


,SEASON,ENDING_DATE,TOURNAMENT,COURSE
0,2025,2025-07-27,3M Open,TPC Twin Cities
1,2024,2024-07-28,3M Open,TPC Twin Cities
2,2023,2023-07-30,3M Open,TPC Twin Cities
3,2022,2022-07-24,3M Open,TPC Twin Cities
4,2021,2021-07-25,3M Open,TPC Twin Cities
5,2020,2020-07-26,3M Open,TPC Twin Cities
6,2019,2019-07-07,3M Open,TPC Twin Cities


### Update Database

#### Old Tournament

In [13]:
importlib.reload(utils.db_utils)  # Only needed if you're actively editing db_utils.py
from utils.db_utils import update_tournament_results

# Change these each year!!
season = 2026
year = 20260  # Unique GraphQL year distinguishing number in case of multiple per year

# Run the update
db_path = "data/golf.db"  # Codex/PGA local copy
tournDf = update_tournament_results(tournament_config, db_path, season, year)

# Show just the most recent tournament added for confirmation
from sqlalchemy import create_engine

engine = create_engine(f"sqlite:///{db_path}")

query = f"""
SELECT *
FROM tournaments
WHERE TOURN_ID = '{tournament_config['old']['id']}'
  AND ENDING_DATE = '{datetime.strptime(tournament_config['old']['date'], '%m/%d/%Y').date()}'
"""

recent = pd.read_sql(query, engine)
engine.dispose()
recent.sort_values(by='FINAL_POS').head()

📦 Fetching results for tournament ID R2026100 (The Open Championship), year: 20260
ℹ️ Tournament 'The Open Championship' already exists — no new data inserted.


,SEASON,ENDING_DATE,TOURN_ID,TOURNAMENT,COURSE,PLAYER,POS,FINAL_POS,ROUNDS:1,ROUNDS:2,ROUNDS:3,ROUNDS:4,OFFICIAL_MONEY,FEDEX_CUP_POINTS
0,2026,2026-07-19,R2026100,The Open Championship,Royal Birkdale GC,Ryan Fox,1,1,+2,-2,-8,-2,"$3,200,000.00",750.000
1,2026,2026-07-19,R2026100,The Open Championship,Royal Birkdale GC,Cameron Young,2,2,-3,-3,+3,-6,"$1,842,000.00",500.000
2,2026,2026-07-19,R2026100,The Open Championship,Royal Birkdale GC,Sam Burns,3,3,+3,-8,-5,+2,"$1,181,000.00",350.000
3,2026,2026-07-19,R2026100,The Open Championship,Royal Birkdale GC,Scottie Scheffler,T4,4,-2,-2,E,-3,"$827,500.00",312.500
4,2026,2026-07-19,R2026100,The Open Championship,Royal Birkdale GC,Tommy Fleetwood,T4,4,-1,-3,-1,-2,"$827,500.00",312.500


#### Last Week's Report Card

Grades the predictions logged before last week's tournament against the results just imported above. The interactive per-player view lives in the app's Prediction Tracker tab.

In [14]:
import importlib, utils.model
importlib.reload(utils.model)
from utils.model import grade_predictions

grade_predictions("data/golf.db")

📋 2 graded week(s) · avg top-15→top-20 hits: 8.00 (eval baseline ≈ 6.5)


,ENDING_DATE,TOURNAMENT,top15_in_top20,expected_top20,top15_cut_rate,status
0,2026-07-19,The Open Championship,8,7.2,0.73,graded
1,2026-07-12,Genesis Scottish Open,8,7.1,0.67,graded


#### Stats

In [15]:
importlib.reload(utils.db_utils)
from utils.db_utils import update_season_stats  # <- This line is essential

# Change these each year!! 
statsYear = 2026

stats_df = update_season_stats(statsYear, db_path)
stats_df.head()

✅ Overwrote stats for season 2026 with 999 rows.


,PLAYER,SGTTG_RANK,SGTTG,SGOTT_RANK,SGOTT,SGAPR_RANK,SGAPR,SGATG_RANK,SGATG,SGP_RANK,...,DRIVING_DISTANCE,DRIVING_ACCURACY_RANK,DRIVING_ACCURACY,GIR_RANK,GIR,SCRAMBLING_RANK,SCRAMBLING,OWGR_RANK,OWGR,SEASON
0,A.J. Ewart,104.0,-0.113,126.0,-0.220,87.0,0.051,64.0,0.056,10.0,...,302.9,94.0,58.39%,36.0,68.70%,12.0,64.79%,183,.8179,2026
1,Aaron Cockerill,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,585,.2149,2026
2,Aaron Rai,22.0,0.744,53.0,0.175,22.0,0.427,46.0,0.141,99.0,...,292.2,2.0,71.70%,14.0,69.78%,107.0,58.09%,20,4.2325,2026
3,Aaron Wise,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,464,.3013,2026
4,Abraham Ancer,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,214,.6880,2026


#### DB Maintenance (via Claude)

Weekly odds are captured automatically: the live-scrape cell in **Current Week** saves each week's odds to the database (`save_current_week_odds`), so season-long imports are no longer part of the routine.

One-off database maintenance is done by asking Claude (in this repo), which uses the helpers kept in `utils/db_utils.py`:

- **Backfill an odds gap** (missed a week, or a whole stretch of events): Claude runs `import_historical_odds(...)` against the golfodds.com archive pages.
- **Rename a player everywhere** (the usual fix — a DK/odds spelling differs from the DB canonical): Claude uses **`consolidate_player_name(db_path, wrong, correct, scope=...)`**. This is the canonical path: it renames the player across **every** table that has a PLAYER column — `odds`, `tournaments`, `stats`, `predictions`, and any table added later (they're discovered automatically, never hard-coded). Pass `scope="dk"` or `scope="player"` to also record the mapping in `name_mappings.json` so the incoming feed self-corrects going forward. It's an exact-match rename, so accented canonicals (Sebastián Muñoz, etc.) are left untouched. **Don't** fix a player one table at a time with `clean_player_names_in_table` — that's how `predictions` silently drifted out of sync (July 2026): a name got corrected in the results tables but not in the prediction log, which broke the Prediction Tracker's results join (Actual Pos showed blank).
- **Check for name drift**: `check_player_name_alignment(db_path)` flags any `predictions` name that matches no results table — the signature of a spelling mismatch. Run it after any DB name edit, or as a routine sanity check. Clean = every logged name lines up.
- **Fix name mismatches in the odds table only** (esp. tournament spelling, or duplicate-spelling collisions): `clean_odds_names(...)` updates the `odds` table in place. For a *player* rename prefer `consolidate_player_name` so the other tables come along too. Never drop/rebuild tables.
- **Course-name drift** (same venue, two spellings): Claude audits and renames to the most recent / PGA API spelling — the History Check cell near the top is the weekly tripwire.
- **Other tripwires**: the `odds coverage` printout in the training build and the per-week ✅/⚠️ checks. If coverage drops, ask Claude to audit and backfill.

The old manual cells (odds name cleanup, yearly archive import, scrape troubleshooting, Misc. Cleaning) were removed July 2026 — they live on in git history if ever needed.

# Model Pipeline (Pooled)

**Redesigned July 2026.** The model now trains on *every* tournament in the database
(point-in-time features, no lookahead) instead of only past editions of this week's
event, and outputs an **expected finish percentile blended with the betting market**
rather than a top-20 probability.

Forward-chained evaluation (2021–2025, 165 events; see `experiments/forward_eval.py`):

| pipeline | hits@15 | AUC |
|---|---|---|
| old per-event method | 5.92 | 0.696 |
| odds-only ranking | 6.38 | 0.722 |
| **this pipeline** | **6.49** | **0.728** |

Key columns in the output:
- **SCORE** — the optimizer input, 1 = best in field. Within-field average of the
  model's rank and the market's rank, rescaled to (0, 1].
- **MODEL_SCORE** — 1 − predicted finish percentile (model only, no market blend).
- **ODDS_SHARE** — vig-normalized implied win probability (the field sums to 1).
- **SG_FORM** — recency-weighted strokes gained per round vs the field
  (100-day halflife), computed from the round scores already in the database.

## Training Data

All events from 2016 through last week, each with features as they stood the
morning of that tournament. Takes ~1 minute. Odds coverage should be ≥95% —
if it drops, run the odds join audit (name mismatches are the usual cause).

In [6]:
import importlib
import utils.features, utils.model
importlib.reload(utils.features)
importlib.reload(utils.model)
from utils.model import build_pooled_training

db_path = "data/golf.db"

training_df, ctx = build_pooled_training(
    db_path,
    first_season=2016,
    as_of=tournament_config["new"]["ending_date"],
)
training_df.tail(3)

✅ 52034 training rows from 401 events (2016–2026), odds coverage 81%


,SEASON,ENDING_DATE,TOURNAMENT,COURSE,PLAYER,POS,FINAL_POS,FINISH_PCT,SGTTG_RANK,SGTTG,...,PCT_FORM_SHRUNK,COURSE_HISTORY,adj_ch,PCT_CH_SHRUNK,ODDS_SHARE,SG_FORM,SG_ROUNDS_12M,SG_CH_SHRUNK,TOP_20,FIELD_SIZE
52031,2026,2026-07-19,The Open Championship,Royal Birkdale GC,Pongsapak Laopakdee,CUT,90,0.576923,NaN,NaN,...,NaN,NaN,NaN,NaN,0.000070,NaN,NaN,NaN,0,156
52032,2026,2026-07-19,The Open Championship,Royal Birkdale GC,Ren Yonezawa,CUT,90,0.576923,NaN,NaN,...,0.4517,NaN,NaN,NaN,0.000701,0.3858,8.0,NaN,0,156
52033,2026,2026-07-19,The Open Championship,Royal Birkdale GC,James Nicholas,W/D,90,0.576923,NaN,NaN,...,0.4833,NaN,NaN,NaN,0.000701,-0.5554,4.0,NaN,0,156


## Current Week

### DraftKings Field

In [7]:
importlib.reload(utils.db_utils)
from utils.db_utils import DK_PLAYER_NAME_MAP
from utils.db_utils import standardize_player_names

# Load DraftKings player list
dk = pd.read_csv("data/DKSalaries.csv", usecols=["Name", "Salary"])

# Normalize DK names to PGA naming convention
dk["Name"] = dk["Name"].replace(DK_PLAYER_NAME_MAP)
dk = dk.rename(columns={"Name": "PLAYER", "Salary": "SALARY"})
dk = standardize_player_names(dk)

print(f"{len(dk)} players in DK field")
dk.head(10)

140 players in DK field


,PLAYER,SALARY
0,Scottie Scheffler,14900
1,Maverick McNealy,9900
2,Kurt Kitayama,9800
3,Hideki Matsuyama,9700
4,Keith Mitchell,9500
5,Jake Knapp,9400
6,Tom Kim,9300
7,Jackson Koivun,9200
8,Pierceson Coody,9100
9,Doug Ghim,9000


### Vegas Odds (live scrape)

Scrapes this week's odds and **saves them into the odds table** so the current
season is fully priced for future training (no more waiting for the year-end
archive).
http://golfodds.com/weekly-odds.html

In [8]:
importlib.reload(utils.db_utils)
from utils.db_utils import get_current_week_odds
from utils.model import save_current_week_odds

odds_current = get_current_week_odds(
    season=tournament_config["new"]["season"],
    tournament_name=new_tournament_name,
)
save_current_week_odds(db_path, odds_current, tournament_config, dry_run=test_mode)
odds_current.head(10)

ℹ️ Odds page is serving: 3M Open (ends 2026-07-26)
ℹ️ This week's odds already saved — nothing to insert.


,SEASON,TOURNAMENT,PLAYER,ODDS,VEGAS_ODDS
1,2026,3M Open,Scottie Scheffler,11/4,2.75
2,2026,3M Open,Maverick McNealy,25/1,25.00
3,2026,3M Open,Kurt Kitayama,25/1,25.00
4,2026,3M Open,Hideki Matsuyama,30/1,30.00
5,2026,3M Open,Tom Kim,30/1,30.00
6,2026,3M Open,Keith Mitchell,30/1,30.00
7,2026,3M Open,Jackson Koivun,30/1,30.00
8,2026,3M Open,Jake Knapp,35/1,35.00
9,2026,3M Open,Corey Conners,40/1,40.00
10,2026,3M Open,Jackson Suber,45/1,45.00


### Feature Rows

Same features as training, computed as of this week: prior-season stats, live
odds → market share, rolling form/cuts/course history, strokes-gained form.

In [9]:
from utils.model import build_current_week_rows

this_week = build_current_week_rows(ctx, dk, odds_current, tournament_config,
                                    allow_stale=test_mode)
this_week.head()

✅ odds coverage: 98% of DK field
⚠️ stats coverage: 68% of DK field
✅ SG form coverage: 97% of DK field
⚠️ No odds AND no SG history (likely name mismatches): ['Jordan L. Smith']


,PLAYER,SALARY,SGTTG_RANK,SGTTG,SGOTT_RANK,SGOTT,SGAPR_RANK,SGAPR,SGATG_RANK,SGATG,...,adj_form,PCT_FORM_SHRUNK,COURSE_HISTORY,adj_ch,PCT_CH_SHRUNK,SG_FORM,SG_ROUNDS_12M,SG_CH_SHRUNK,ODDS_SHARE,FIELD_SIZE
0,Scottie Scheffler,14900,1.0,2.361,2.0,0.748,1.0,1.291,16.0,0.322,...,4.19,0.1889,NaN,NaN,NaN,2.2571,78.0,NaN,0.188611,140
1,Maverick McNealy,9900,54.0,0.390,61.0,0.185,48.0,0.280,120.0,-0.075,...,11.85,0.3701,39.5,24.54,0.3359,0.7324,80.0,1.0024,0.027204,140
2,Kurt Kitayama,9800,7.0,1.094,12.0,0.535,32.0,0.389,39.0,0.170,...,11.75,0.3441,3.5,3.19,0.2612,0.6569,93.0,2.3928,0.027204,140
3,Hideki Matsuyama,9700,22.0,0.789,139.0,-0.171,19.0,0.483,5.0,0.477,...,8.73,0.3229,30.0,43.28,0.3974,0.6264,88.0,0.0532,0.022816,140
4,Keith Mitchell,9500,45.0,0.513,13.0,0.533,65.0,0.172,151.0,-0.193,...,13.78,0.3821,47.2,26.34,0.3595,0.6722,92.0,1.0728,0.022816,140


### Data Health Checks

A DK player missing **both** odds and SG history is almost always a name-map
miss (fix in `DK_PLAYER_NAME_MAP` / `PLAYER_NAME_MAP` in `utils/db_utils.py`),
not a real unknown. Missing one of the two is normal for rookies/alternates.

In [10]:
check = this_week[["PLAYER", "SALARY", "VEGAS_ODDS", "SG_FORM", "SGTTG"]].copy()
check["n_missing"] = check[["VEGAS_ODDS", "SG_FORM", "SGTTG"]].isna().sum(axis=1)
suspects = check[check["n_missing"] >= 2].sort_values("SALARY", ascending=False)
print(f"{len(suspects)} players missing 2+ key features:")
display(suspects)

# High-salary players with no odds match: usually an odds-side name typo
# (e.g. golfodds' "Piercescon Coody"). Fix with scope="player" below.
no_odds = check[check["VEGAS_ODDS"].isna() & (check["SALARY"] >= 7000)]
if len(no_odds):
    print()
    print(f"{len(no_odds)} players at $7k+ with no odds match (check the odds board for typos):")
    display(no_odds.sort_values("SALARY", ascending=False))

6 players missing 2+ key features:


,PLAYER,SALARY,VEGAS_ODDS,SG_FORM,SGTTG,n_missing
16,Jordan L. Smith,8300,NaN,NaN,NaN,3
19,Benjamin James,8000,NaN,0.0400,NaN,2
93,S.Y. Noh,6600,NaN,-0.2916,NaN,2
104,Stefano Mazzoli,6500,400.0,NaN,NaN,2
137,Sihan Sandhu,6000,5000.0,NaN,NaN,2
139,Muzzy Donohue,6000,5000.0,NaN,NaN,2



2 players at $7k+ with no odds match (check the odds board for typos):


,PLAYER,SALARY,VEGAS_ODDS,SG_FORM,SGTTG,n_missing
16,Jordan L. Smith,8300,NaN,NaN,NaN,3
19,Benjamin James,8000,NaN,0.04,NaN,2


### Fix Name Mismatches

Add mappings here — no need to edit `utils/db_utils.py`. They persist to
`data/name_mappings.json` and apply on every future run.

- `scope="dk"` — the DraftKings salary file spells it differently than the DB
  (e.g. `add_name_mapping("John Keefer", "Johnny Keefer", scope="dk")`)
- `scope="player"` — the odds scrape / PGA results spell it differently
  (e.g. `add_name_mapping("Johnny Keefer", "John Keefer")`)

The first argument is the WRONG (incoming) spelling, the second is the DB's
canonical spelling. After adding, re-run from the DraftKings Field cell.

In [10]:
from utils.db_utils import add_name_mapping, show_name_mappings

# "Wrong Name", "Correct Name"

# add_name_mapping("Matthew McCarty", "Matt McCarty", scope="dk")           # DK file side
# add_name_mapping("Wrong Name", "Canonical DB Name", scope="player")      # odds/results side
# add_name_mapping("Old Tourney Name", "Canonical Tourney Name", scope="tournament")    # odds/results side

# show_name_mappings()  # uncomment to print the full list of every mapping

### Column Glossary (derived features)

All rolling windows end the day **before** the tournament's ending date (equivalent
to its start morning — no other event finishes mid-week). "Percentile" always means
finish position within that event's field: **0 = won, 1 = finished last**; missed
cuts land together in the cut bucket (≈0.58 in a 156-man field).

**Market**
- `VEGAS_ODDS` — scraped fractional odds as a decimal (30/1 → 30.0). Display only; not a model input.
- `ODDS_SHARE` — the model's market feature. Implied win probability `1/(odds+1)`;
  unlisted players get half the field's smallest listed probability; then everything
  is divided by the field total so shares **sum to exactly 1**. Read it as the
  player's percent of the field's win equity (0.047 = 4.7%).

**Form**
- `SG_FORM` — strokes per round better (+) or worse (−) than the field average,
  built from every round in the last 2 years. A round's weight halves every 100
  days (a round from ~3 months ago counts half; ~6 months ago a quarter). Thin
  samples are shrunk toward 0 (field average). Guide: +1.5 elite-hot, +0.5 solid,
  0 tour-average, −1 struggling fringe. Missing (no rounds in 2 yrs) → filled at
  the training set's 25th percentile at fit time ("assume below average").
- `SG_ROUNDS_12M` — competitive rounds in the last 365 days; the sample size
  behind SG_FORM (82 = trust it, 6 = don't).
- `PCT_FORM_SHRUNK` — average finish percentile over the last 9 months, blended
  with **4 phantom perfectly-average (0.5) events** so small samples pull toward
  the middle. 0.37 = "typically finishes in the top ~37% of fields, sample-adjusted."
  Lower = better. Missing (inactive) → 0.8.
- `CUT_PERCENTAGE` — % of cuts made in the last 9 months.
- `CONSECUTIVE_CUTS` — current made-cut streak entering this week.
- `FEDEX_CUP_POINTS` — points earned in the last 9 months (rolling window, NOT the
  season-to-date total).
- `form_density` — FedEx points per event in that window.

**Course**
- `SG_CH_SHRUNK` — strokes gained per round AT this course over the past 7 years,
  shrunk toward 0 (field average) by 2 phantom average rounds. +1.5 = proven course
  horse, 0 = no history / average, negative = struggles here. Replaced the
  finish-position-based PCT_CH_SHRUNK (Stage 6: beat it in 4 of 5 eval seasons —
  strokes are a cleaner signal than finish positions).

**Context**
- `FIELD_SIZE` — number of players in the event. Calibration context for the model
  (what a given ODDS_SHARE means, how big the cut bucket is). Identical for every
  player in a given week, so it cancels out of the within-field ranking.

**Prior-season stats block** (`SGTTG`, `SGOTT`, `SGAPR`, `SGATG`, `SGP`, driving,
GIR, scrambling, OWGR + their `_RANK`s) — last season's full-year PGA Tour
aggregates. Individually near-zero importance (stale + redundant with odds/SG
form) but collectively still worth a small edge; kept as model inputs, mostly
dropped from the export.

## Train + Predict

A random-forest regressor is trained on every historical row (~50k) to predict
`FINISH_PCT`, then this week's field is scored and blended with the market.

**`MODEL_SCORE` = 1 − predicted finish percentile.** Pure model output (odds are
one of its inputs, but this number is whatever the forest says). 0.76 = "expected
to finish around the 24th percentile of this field." Magnitudes are meaningful —
a gap of 0.04 is a real skill gap, not just a rank step.

**`SCORE` — the optimizer column — is computed within this week's field:**

1. Rank all N players by `MODEL_SCORE` (1 = worst … N = best).
2. Rank all N players by `ODDS_SHARE`. Bookmakers quote tied prices (all the
   55/1s share one averaged rank) — the model is what breaks those ties.
3. Average the two ranks: `blended_rank = (model_rank + market_rank) / 2`.
4. Rescale: `SCORE = blended_rank / N` → range (0, 1], **1.0 = best in field**.

It's a blend of *ranks*, not values, because the two live on incomparable scales
(MODEL_SCORE ~0.4–0.8 vs ODDS_SHARE ~0.0002–0.05). The 50/50 weight is validated:
forward-chained eval scored the blend 6.49 hits@15 vs 6.41 model-only and 6.38
odds-only, with a flat optimum anywhere from 25% to 85% market weight.

**`LEVERAGE` = model_rank − market_rank.** Positive = the model likes the player
more than the market does (potential value play); negative = the market is higher
on him than the model. The biggest absolute values are this week's disagreements —
worth a look before locking a lineup.

In [11]:
from utils.model import train_and_score

scored, importances = train_and_score(training_df, this_week)

preview_cols = ["PLAYER", "SALARY", "P_TOP20", "SCORE", "MODEL_SCORE", "ODDS_SHARE", "LEVERAGE",
                "VEGAS_ODDS", "SG_FORM", "PCT_FORM_SHRUNK", "SG_CH_SHRUNK",
                "CUT_PERCENTAGE"]
scored[[c for c in preview_cols if c in scored.columns]].head(30)

c:\Users\ryanh\Documents\Github\pga-dk\utils\features.py:298: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  f["PCT_CH_SHRUNK"] = f["PCT_CH_SHRUNK"].fillna(0.5)
c:\Users\ryanh\Documents\Github\pga-dk\utils\features.py:305: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  f["SG_CH_SHRUNK"] = f["SG_CH_SHRUNK"].fillna(0.0)
c:\Users\ryanh\Documents\Github\pga-dk\utils\features.py:311: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-i

,PLAYER,SALARY,P_TOP20,SCORE,MODEL_SCORE,ODDS_SHARE,LEVERAGE,VEGAS_ODDS,SG_FORM,PCT_FORM_SHRUNK,SG_CH_SHRUNK,CUT_PERCENTAGE
0,Scottie Scheffler,13300,0.7779,1.0000,0.853066,0.083698,0.0,7.5,2.2475,0.1970,0.0,93.8
1,Rory McIlroy,11900,0.6632,0.9936,0.808058,0.074888,0.0,8.5,1.7451,0.2549,0.0,100.0
2,Tommy Fleetwood,10500,0.5981,0.9856,0.792739,0.041849,0.5,16.0,1.4049,0.2847,0.0,93.3
3,Matt Fitzpatrick,10000,0.5553,0.9824,0.786005,0.041849,-0.5,16.0,1.5598,0.2665,0.0,100.0
4,Ludvig Aberg,9300,0.4482,0.9407,0.720136,0.022950,-3.5,30.0,1.0639,0.3163,0.0,86.7
5,Collin Morikawa,8700,0.4380,0.9503,0.724384,0.022950,-0.5,30.0,1.2483,0.3160,0.0,91.7
6,Wyndham Clark,8200,0.4380,0.9455,0.729663,0.019762,7.0,35.0,1.4328,0.3572,0.0,78.9
7,Robert MacIntyre,8000,0.4354,0.9567,0.726526,0.022950,1.5,30.0,0.9208,0.3972,0.0,83.3
8,Xander Schauffele,9800,0.4303,0.9535,0.721745,0.027363,-5.5,25.0,0.8862,0.3447,0.0,86.7
9,Viktor Hovland,8300,0.4251,0.9231,0.715445,0.019762,0.0,35.0,1.0144,0.3826,0.0,80.0


In [12]:
import plotly.express as px

imp_df = importances.head(20).reset_index()
imp_df.columns = ["Feature", "Importance"]
fig = px.bar(imp_df, x="Feature", y="Importance",
             title="Feature Importances (Percentile Regressor)",
             template="plotly_dark")
fig.update_layout(xaxis_tickangle=-45)
fig.show()

## Export

Sorted by **P_TOP20** — the key column: a calibrated probability of finishing
top-20 (model blended with market-implied, magnitudes learned from a decade of
odds-vs-results). Point the Excel optimizer at it: **the lineup's sum of
P_TOP20 = its expected number of top-20 finishers.**

SCORE (rank blend, 1 = best) is kept as the pure-ordering view; LEVERAGE
(model rank − market rank) remains the value/fade tilt.


In [ ]:
export_cols = ["PLAYER", "SALARY", "P_TOP20", "SCORE", "MODEL_SCORE", "ODDS_SHARE", "LEVERAGE",
               "VEGAS_ODDS", "SG_FORM",
               "PCT_FORM_SHRUNK", "SG_CH_SHRUNK",
               "CUT_PERCENTAGE", "FEDEX_CUP_POINTS", "OWGR_RANK"]

export_df = scored[[c for c in export_cols if c in scored.columns]].copy()

round_map = {"P_TOP20": 3, "MODEL_SCORE": 4, "ODDS_SHARE": 4, "LEVERAGE": 1, "SG_FORM": 2,
             "PCT_FORM_SHRUNK": 3, "SG_CH_SHRUNK": 2,
             "CUT_PERCENTAGE": 1, "FEDEX_CUP_POINTS": 0, "OWGR_RANK": 0}
for col, nd in round_map.items():
    if col in export_df.columns:
        export_df[col] = export_df[col].round(nd)

export_df = export_df.sort_values("P_TOP20", ascending=False).reset_index(drop=True)
export_df.to_csv("data/current_week_export.csv", index=False)
print(f"✅ Exported {len(export_df)} players to data/current_week_export.csv")


# Log this week's forecast to the predictions table (feeds the Report Card cell
# and the app's Prediction Tracker; re-running the same week is a no-op)
from utils.model import save_predictions, save_current_week_meta
save_predictions("data/golf.db", export_df, tournament_config, dry_run=test_mode)

# Refresh the app's current-week marker (data/current_week.json) so the Course
# Explorer defaults to THIS week's course and the As-of date defaults to this
# week. Skipped in test_mode, whose config points at a past week — writing it
# then would leave the app defaulting to a stale course.
if not test_mode:
    _meta = save_current_week_meta(tournament_config)
    print(f"✅ App marker → {_meta['name']} @ {_meta['course']} ({_meta['ending_date']})")

export_df.head(15)